### ISSUE TITLE
Cyclic references could lead to memory leaks

### ISSUE THREAD

**Issue by Thenerdstation** [google/jax#725]
Cyclic references (for example, in graph data structures) don't get garbage collected by python immediately. This can cause tensors to not be deallocated when all references are removed, causing memory leaks on device and possible OOM crashes.

**Comment by jekbradbury** [comment]
Can you give a MWE of what you mean by in-graph data structures? You might be able to fix this with Python weakrefs if the cycles are created in data structures your code manages.

**Comment by Thenerdstation** [comment]
Yeah that's what we ended up doing. But weakref is a fairly low level API to be using.

**Comment by hawkinsp** [comment]
Can you clarify what you mean? Is your concern that (a) references don't get collected by Python *immediately* or (b) that references don't get collected at all?

I expect you mean the former, in which case there isn't a whole lot we can do about it at our level — it's something you'll have to fix in your code. The only thing I can possibly suggest is that we could add a Delete() method to our DeviceArrays to give you more explicit control over destruction.

I'd be surprised if the latter is happening, but if you have a repro I'd be happy to look into it.

**Comment by mattjj** [comment]
Put another way: is this an issue with JAX specifically, or would e.g. regular NumPy have this problem?

**Comment by Thenerdstation** [comment]
Yes adding a `Delete` method would be great. 

The issue is amplified with JAX when compared to numpy since python calls garbage collection when it starts to run out of runtime memory, but not when JAX runs out if device memory.

Note this cyclic issue is only true outside of a `jit`.
